### DISCUSSION TITLE
View() object no longer available

### DISCUSSION THREAD

**Discussion by brndngrhm** [posit-dev/positron#4658] [Q&A]
i often use `View()` at the end of dplyr chains to quickly check my logic before assigning the output to an object.

In Positron, I'm finding that once i execute a 2nd `View()` command, the first one is no longer viewable (see below). I am used to being lazy and just comparing outputs across multiple View() statements without having to assign to objects, which i am able to do in R Studio but not in Positron.

Is this capability something that can be added to Positron?

```
library(tidyverse)

# View 1 - executing this opens the data pane showing mtcars (screenshot 1)
mtcars %>% View()

# View 2 - executing this opens a new data pane showing summarized mtcars  (screenshot 2)
#        - trying to go back to the first data pane results in a message: object no longer available (screenshot 3)
mtcars %>% 
  group_by(mpg) %>% 
  tally() %>% 
  View()
```

**Screenshot 1:**

<img width="1075" alt="Screenshot 2024-09-12 at 9 11 41 AM" src="https://github.com/user-attachments/assets/4ecdf87f-0731-4d36-a7b8-4d31f539784c">

**Screenshot 2:**

<img width="724" alt="Screenshot 2024-09-12 at 9 12 44 AM" src="https://github.com/user-attachments/assets/b289f914-33e8-4bde-9250-0e220440d287">

**Screenshot 3:**

<img width="722" alt="Screenshot 2024-09-12 at 9 13 07 AM" src="https://github.com/user-attachments/assets/05e655a9-d648-406c-b77f-e51b3853b7c7">

**Comment by jennybc** [comment]
@jthomasmock Can you comment on whether this is intended behaviour?

**Reply by jmcphers** [reply]
I think it's "sort of" intended. RStudio's data viewer makes a copy of everything you give it, so data viewer tabs can keep working if you remove the original copy. Positron's data viewer doesn't ever make copies; it only runs against the original object. A simpler repo case:

```r
mt <- mtcars
View(mt)
rm(mt)
```

In Positron this will result in your data viewer immediately going offline with the "object is no longer available" message, whereas in RStudio you can keep using the viewer.

Choosing to serve all requests from the original data (vs making a cached copy) was an intentional design decision that was made with the goal of keeping memory use under control (especially for big objects). But we could walk it back or fine tune it, especially for R (where copy-on-write can make most copies zero/low cost) and when the data is small.

**Reply by jthomasmock** [reply]
Yup, agreed with Jonathan. 😄 

I do think it's worth opening as an issue @brndngrhm so we can track the request proper. I'll convert your question into an issue with Jonathan's response.

**Reply by jthomasmock** [reply]
Issue up at: https://github.com/posit-dev/positron/issues/4668

**Reply by brndngrhm** [reply]
thanks!
